cypress/e2e/shared/notificationRules.test.ts

```diff
@@ -292,4 +302,457 @@ describe('NotificationRules', () => {
+    const endp: GenEndpoint = {
+      orgID: '',
+      name: 'Test HTTP Endpoint',
+      userID: '',
+      description: 'A Dummy Endpoint in AWS Land',
+      status: 'active',
+      type: 'http',
+      url: 'http://endpoint.devnull.io:3000',
+      authMethod: 'none',
+      method: 'POST',
+    }
+
+    const rule: GenRule & {activeStatus: string} = {
+      type: 'http',
+      every: '1m',
+      offset: '0m',
+      url: '',
+      orgID: '',
+      name: 'Test Rule',
+      activeStatus: 'active',
+      status: 'active',
+      endpointID: '',
+      tagRules: [],
+      labels: [],
+      statusRules: [{currentLevel: 'CRIT', period: '1h', count: 1}],
+      description: '',
+    }
+
+    const check: GenCheck = {
+      type: 'deadman',
+      name: 'Ghost Check',
+      status: 'active',
+      orgID: '',
+      query: {
+        name: '',
+        text: '',
+        editMode: 'builder',
+        builderConfig: {
+          buckets: [],
+          tags: [
+            {
+              key: '_measurement',
+              values: ['wumpus'],
+              aggregateFunctionType: 'filter',
+            },
+            {key: '_field', values: ['mag'], aggregateFunctionType: 'filter'},
+            {key: 'foo', values: [], aggregateFunctionType: 'filter'},
+          ],
+          functions: [],
+        },
+        //        hidden: false
+      },
+      labels: [],
+      every: '1m',
+      level: 'CRIT',
+      offset: '0s',
+      reportZero: false,
+      staleTime: '10m',
+      statusMessageTemplate: 'Check: ${ r._check_name } is: ${ r._level }',
+      tags: [],
+      timeSince: '90s',
+    }
+
+    interface MonitoringRec {
+      check: GenCheck
+      endp: NotificationEndpoint
+      rule: GenRule
+      level: string
+      source: string
+      sent: string
+      valField: {key: string; val: string}
+      sourceTimestamp: number
+      statusTimestamp: number
+      message: string
+    }
+
+    const writeMonitoringRecord = (rec: MonitoringRec, offset: string) => {
+      const lp =
+        `notifications,_level=${rec.level},_notification_endpoint_id=${rec.endp.id},` +
+        `_notification_endpoint_name=${rec.endp.name.replace(
+          / /g,
+          '\\ '
+        )},_notification_rule_id=${rec.rule.id},` +
+        `_notification_rule_name=${rec.rule.name.replace(/ /g, '\\ ')},` +
+        `_check_id=${rec.check.id},_check_name=${rec.check.name.replace(
+          / /g,
+          '\\ '
+        )},` +
+        `_sent=${rec.sent},_source_measurement=${rec.source},_type=deadman,${
+          rec.valField.key
+        }=${rec.valField.val.replace(/ /g, '\\ ')} ` + // end tags start fields
+        `_source_timestamp=${rec.sourceTimestamp},_status_timestamp=${
+          rec.statusTimestamp
+        },_message="${rec.message.replace(/ /g, ' ')}",dead="true"`
+
+      cy.writeLPData({lines: [lp], offset: offset, namedBucket: '_monitoring'})
+    }
+
+    beforeEach('Generate alert records', () => {
```

Comment:
I've seen flaky tests in the past from `beforeEach` functions that don't handle async functions & promises. I think we should refactor this to use cypress chaining (`cy.xyz().then(() => cy.abc().then(() => ...`) to make sure the beforeEach doesn't finish prematurely.

Reply:
Generalized this to a `localInit()` method that returns CypressChainable.  It's at the start of the chain.  It handles a lot of local data.  It is probably too specific to warrant being generalized into a command.

Reply:
Cool, looks much better. You could move it back into a `beforeEach` to avoid having to call it in each test. As long as the code within the `beforeEach` returns a Cypress chainable object, it will wait for completion before starting the test. But this implementation is fine too.